core-bundle/src/Resources/contao/controllers/FrontendIndex.php

```diff
@@ -263,23 +265,23 @@ public function renderPage($pageModel)
 		// Authenticate the user if the page is protected
 		if ($objPage->protected)
 		{
-			$container = System::getContainer();
-			$token = $container->get('security.token_storage')->getToken();
+			$security = System::getContainer()->get('security.helper');
 
-			if ($container->get('security.authentication.trust_resolver')->isAnonymous($token))
+			if (!$security->isGranted(ContaoCorePermissions::MEMBER_IN_GROUPS, $objPage->groups))
 			{
-				throw new InsufficientAuthenticationException('Not authenticated: ' . Environment::get('uri'));
-			}
+				if ($security->getToken() instanceof AnonymousToken)
+				{
+					throw new InsufficientAuthenticationException('Not authenticated: ' . Environment::get('uri'));
+				}
 
-			if (!$container->get('security.authorization_checker')->isGranted('ROLE_MEMBER'))
-			{
-				throw new AccessDeniedException('Access denied: ' . Environment::get('uri'));
-			}
+				$user = $security->getUser();
 
-			// Check the user groups
-			if (!$this->User->isMemberOf($objPage->groups))
-			{
-				$this->log('Page ID "' . $objPage->id . '" can only be accessed by groups "' . implode(', ', (array) $objPage->groups) . '" (current user groups: ' . implode(', ', $this->User->groups) . ')', __METHOD__, TL_ERROR);
+				if (!$user instanceof FrontendUser)
+				{
+					throw new AccessDeniedException('Access denied: ' . Environment::get('uri'));
+				}
+
+				$this->log('Page ID "' . $objPage->id . '" can only be accessed by groups "' . implode(', ', $objPage->groups) . '" (current user groups: ' . implode(', ', StringUtil::deserialize($user->groups, true)) . ')', __METHOD__, TL_ERROR);
 
 				throw new AccessDeniedException('Access denied: ' . Environment::get('uri'));
```

Comment:
```suggestion
				if ($user instanceof FrontendUser)
				{
					$this->log('Page ID "' . $objPage->id . '" can only be accessed by groups "' . implode(', ', $objPage->groups) . '" (current user groups: ' . implode(', ', StringUtil::deserialize($user->groups, true)) . ')', __METHOD__, TL_ERROR);
				}

				throw new AccessDeniedException('Access denied: ' . Environment::get('uri'));
```

Would remove the need for the duplicated throw ☺️

Reply:
Changed in 7322e53160c58bab70644a88f6239dbfa8474d08